plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Internal.hs

```diff
@@ -0,0 +1,121 @@
+import           Ide.Plugin.Error                     (PluginError (PluginInternalError, PluginRuleFailed),
+                                                       getNormalizedFilePathE)
+import           Ide.Plugin.SemanticTokens.Mappings
+import           Ide.Plugin.SemanticTokens.Query
+import           Ide.Plugin.SemanticTokens.Types
+import           Ide.Types
+import qualified Language.LSP.Protocol.Lens           as L
+import           Language.LSP.Protocol.Message        (Method (Method_TextDocumentSemanticTokensFull))
+import           Language.LSP.Protocol.Types          (NormalizedFilePath,
+                                                       SemanticTokens,
+                                                       type (|?) (InL))
+import           Prelude                              hiding (span)
+
+logWith :: (MonadIO m) => IdeState -> Priority -> String -> m ()
+logWith st prior = liftIO . logPriority (ideLogger st) prior . T.pack
+
+-----------------------
+---- the api
+-----------------------
+
+computeSemanticTokens :: IdeState -> NormalizedFilePath -> ExceptT PluginError Action SemanticTokens
+computeSemanticTokens state nfp = do
+    let dbg = logWith state Debug
+    dbg $ "Computing semantic tokens for: " <> show nfp
+    (RangeHsSemanticTokenTypes {tokens}, mapping) <- useWithStaleE GetSemanticTokens nfp
+    let rangeTokens = mapMaybe (\(span, name) -> (,name) <$> toCurrentRange mapping span) tokens
+    withExceptT PluginInternalError $ liftEither $ semanticTokenAbsoluteSemanticTokens rangeTokens
+
+semanticTokensFull :: PluginMethodHandler IdeState 'Method_TextDocumentSemanticTokensFull
+semanticTokensFull state _ param = do
+  nfp <- getNormalizedFilePathE (param ^. L.textDocument . L.uri)
+  items <- runActionE "SemanticTokens.semanticTokensFull" state $ computeSemanticTokens state nfp
+  return $ InL items
+
+-- | Defines the 'getSemanticTokensRule' function, compute semantic tokens for a Haskell source file.
+--
+-- This Rule collects information from various sources, including:
+--
+-- Imported name token type from Rule 'GetSemanticTokens'
+-- Local names token type from 'hieAst'
+-- Name locations from 'hieAst'
+-- Visible names from 'tmrRenamed'
+--
+-- It then combines this information to compute the semantic tokens for the file.
+--
+getSemanticTokensRule :: Recorder (WithPriority Log) -> Rules ()
+getSemanticTokensRule recorder =
+  define (cmapWithPrio LogShake recorder) $ \GetSemanticTokens nfp -> do
+    (hscEnv -> hsc) <- use_ GhcSessionDeps nfp
+    (HAR {..}) <- use_ GetHieAst nfp
+    Just (_, ast) <- return $ listToMaybe $ Map.toList $ getAsts hieAst
+    -- get current location from the old ones
+    let spanNamesMap = hieAstSpanNames ast
+    let nameSet = unionNameSets $ Map.elems spanNamesMap
+    -- get imported name semantic map
+    importedNameSemanticMap <- liftIO $ foldrM (getTypeExclude hsc) emptyNameEnv $ nameSetElemsStable nameSet
```

Comment:
why do we need to consult the `hsc` environment for the type? Can't we get them from the hie ast or the refmap?

Reply:
we currently use tyThing to ease the implementation, since GHC provide a handful of functions to distinguish the token semantic. Not sure about how to recover tyThing  from hie ast or the refmap. 
Or Is there a better way than using tyThing?

Reply:
For either way, please enlighten me, I'll take a look at them latter.

Reply:
why can't we use `typeSemantic` instead of `tyThingSemantic`?

Reply:
I tried, but I failed find enough semantic information from HieTypeFix, such as class method, pattern synonym etc..

Reply:
what about in the `OccName` and its `NameSpace`?

Reply:
no luck here too.